spam/03_spam_data_slicing_tutorial.py

```diff
@@ -362,80 +249,198 @@ def textblob_polarity(x):
+# * `input_data_key`: corresponds to the desired input field from the `X_dict`
+# * `task_name`: corresponds to the corresponding input field from the `Y_dict`
+
+# %%
+from snorkel.slicing import BinarySlicingClassifier
+from utils import get_pytorch_mlp
+
+
+# Define model architecture
+bow_dim = X_train.shape[1]
+mlp = get_pytorch_mlp(hidden_dim=bow_dim, num_layers=2)
+
+# Init slice model
+slice_model = BinarySlicingClassifier(
+    base_architecture=mlp,
+    head_dim=bow_dim,
+    input_data_key="bow_features",
+    task_name="spam_task",
+    slice_names=[sf.name for sf in sfs],
+)
+
+# %% [markdown]
+# ### Monitor slice performance _during training_
+#
+# Using Snorkel's [`Trainer`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.Trainer.html), we fit to `train_dl`, and validate on `valid_dl`.
+#
+# We note that we can monitor slice-specific performance during training — this is a powerful way to track especially critical subsets of the data.
+# If logging in `Tensorboard` (i.e. [`snorkel.classification.TensorboardWritier`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.TensorBoardWriter.html)), we would visualize individual loss curves and validation metrics to debug convegence for specific slices.
+
+# %%
+from snorkel.classification import Trainer
+
+# For demonstration purposes, we set n_epochs=2
+trainer = Trainer(lr=1e-4, n_epochs=2)
+trainer.fit(slice_model, [train_dl, valid_dl])
+
+# %% [markdown]
+# ### Slice-specific evaluation with `BinarySliceClassifier`
+
+# %% [markdown]
+# With the `score` method, we'd like to evaluate _overall performance_.
+
+# %%
+slice_model.score([valid_dl], as_dataframe=True)
+
+# %% [markdown]
+# In overall metrics (e.g. f1, accuracy) our model appears to perform well!
+# However, we might actually be more interested in performance for application-critical subsets, as defined by slices.
+
+# %% [markdown]
+# In order to evaluate on fine-grained slices, we'd like to initialize a **slice-aware dataloader**.
+# To do this, we can use [`slice_model.make_slice_dataloader`]() to add slice labels to an existing dataloader.
+#
+# Under the hood, this method leverages slice metadata to add slice labels to the appropriate fields such that it's compatible with the initialized `SliceBinaryClassifier`.
+
+# %%
+S_train = applier.apply(df_train)
+S_valid = applier.apply(df_valid)
+
+valid_dl_slice = slice_model.make_slice_dataloader(valid_dl.dataset, S_valid)
+test_dl_slice = slice_model.make_slice_dataloader(test_dl.dataset, S_test)
+
+# %% [markdown]
+# Then, we can use the [`score_slices`]() in `BinarySliceClassifier` to natively monitor performance by slice.
+#
+# *NOTE:* This method differs from `score` — it remaps all slice-related labels, denoted `spam_task_slice:{slice_name}_pred`, to be evaluated on the `spam_task`.
+
+# %%
+slice_model.score_slices([valid_dl_slice, test_dl_slice], as_dataframe=True)
```

Comment:
All the above feels redundant with the previous section on monitoring.  I'd delete and just focus on training the slice aware model